feat(bootstrap): add platform and app initializers

vsavkin opened the pull request:
Often some init logic needs to run when a platform or an application is boostrapped.
For example, boostraping a platform requires initializing the dom adapter.
Now, it can be done as follows:

new Provider(PLATFORM_INITIALIZER, {useValue: initDomAdapter, multi: true}),

All platform initializers will be run after the platform injector has been created.

Similarly, all application initializers will be run after the app injector has been
created.

IgorMinar commented on modules/angular2/src/core/application_ref.ts:
```diff
@@ -439,7 +465,7 @@ export class ApplicationRef_ extends ApplicationRef {
   dispose(): void {
     // TODO(alxhub): Dispose of the NgZone.
-    this._rootComponents.forEach((ref) => ref.dispose());
+    ListWrapper.clone(this._rootComponents).forEach((ref) => ref.dispose());
```
why clone?

vsavkin replied:
Because we are removing items from the collection while iterating over it. This never worked.

IgorMinar replied:
lol

On Wed, Nov 18, 2015 at 8:35 AM Victor Savkin <notifications@github.com>
wrote:

> In modules/angular2/src/core/application_ref.ts
> <https://github.com/angular/angular/pull/5355#discussion_r45223098>:
>
> > @@ -439,7 +465,7 @@ export class ApplicationRef_ extends ApplicationRef {
> >
> >    dispose(): void {
> >      // TODO(alxhub): Dispose of the NgZone.
> > -    this._rootComponents.forEach((ref) => ref.dispose());
> > +    ListWrapper.clone(this._rootComponents).forEach((ref) => ref.dispose());
>
> Because we are removing items from the collection while iterating over it.
> This never worked.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular/pull/5355/files#r45223098>.
>

IgorMinar commented on modules/angular2/src/platform/browser_common.ts:
```diff
@@ -11,13 +11,13 @@ import {
 import {Testability} from 'angular2/src/core/testability/testability';
 
-// TODO change these imports once dom_adapter is moved out of core
 import {DOM} from 'angular2/src/core/dom/dom_adapter';
```
so DOM remains in the core, but the actual implementation is provided by a platform barrel?

vsavkin replied:
Yeah, next step is moving it out.

IgorMinar commented:
overall this looks good to me. please answer the questions above and merge.

IgorMinar commented:
and get CI to pass

vsavkin committed:
feat(bootstrap): add platform and app initializers

Often some init logic needs to run when a platform or an application is boostrapped.
For example, boostraping a platform requires initializing the dom adapter.
Now, it can be done as follows:

new Provider(PLATFORM_INITIALIZER, {useValue: initDomAdapter, multi: true}),

All platform initializers will be run after the platform injector has been created.

Similarly, all application initializers will be run after the app injector has been
created.
